### PR TITLE
RFC: shooter subsystem and commands

### PR DESCRIPTION
**Pull request by duane701**
RFC -- Request for Comment

Starting a PR to make comments about the shooter

### CODE REVIEW

File: src/Robot.h [hunk 1]
```diff
@@ -32,7 +34,9 @@ class Robot : public IterativeRobot {
 	// BEGIN AUTOGENERATED CODE, SOURCE=ROBOTBUILDER ID=DECLARATIONS
     static std::shared_ptr<Chassis> chassis;
 	// END AUTOGENERATED CODE, SOURCE=ROBOTBUILDER ID=DECLARATIONS
-	virtual void RobotInit();
+    static std::shared_ptr<Chassis> chassis;
```
Comment (duane701) [inline src/Robot.h]: Copy pasta? This line is the same as the line above. Probably should be instantiating the shooter subsystem.

File: src/Subsystems/Shooter.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#include "Shooter.h"
+#include "../RobotMap.h"
+
+Shooter::Shooter() :
+		Subsystem("Shooter")
```
Comment (duane701) [inline src/Subsystems/Shooter.cpp]: Please set Tabs to Spaces. 2 spaces per tab.

File: src/Subsystems/Shooter.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+{
+	leftFlywheel = RobotMap::shooterLeftFlywheel;
+	rightFlywheel = RobotMap::shooterRightFlywheel;
+	roller = RobotMap::shooterRoller;
+
+	lifter = RobotMap::shooterLifter;
```
Comment (duane701) [inline src/Subsystems/Shooter.cpp]: We should learn how to use the initializer list. This constructor should look like:

```
Shooter::Shooter() :
  Subsystem("Shooter"),
  leftFlywheel(RobotMap::shooterLeftFlywheel),
  rightFlywheel(RobotMap::shooterRightFlywheel),
  roller(RobotMap::shooterRoller),
  lifter(RobotMap::shooterLifter)
{
  rightFlywheel->SetInverted(true);
}
```

File: src/Subsystems/Shooter.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+	rightFlywheel->SetInverted(true);
+}
+
+void Shooter::InitDefaultCommand()
+{
+	// Set the default command for a subsystem here.
+	//SetDefaultCommand(new MySpecialCommand());
```
Comment (duane701) [inline src/Subsystems/Shooter.cpp]: Remove unused code.

File: src/Subsystems/Shooter.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// here. Call these from Commands.
+
+
+
+void Shooter::SetLifter(LifterValue value)
+{
+	lifter->Set(static_cast<DoubleSolenoid::Value>(value));
+}
+
+bool Shooter::IsLifterHigh()
```
Comment (duane701) [inline src/Subsystems/Shooter.cpp]: Change `IsLifterHigh` to `IsLifterUp` to better match the enum names.

File: src/Subsystems/Shooter.cpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+{
+	return lifter->Get() == DoubleSolenoid::kForward;
+}
+
+void Shooter::SetShooter(double speed)
+{
+	leftFlywheel->Set(speed);
+	rightFlywheel->Set(speed);
```
Comment (duane701) [inline src/Subsystems/Shooter.cpp:41]: Can we set `rightFlyWheel` to follow?

File: src/Subsystems/Shooter.h [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+	// for methods that implement subsystem capabilities
+public:
+	Shooter();
+	void InitDefaultCommand();
+
+	enum LifterValue {kUp = DoubleSolenoid::kForward, kDown = DoubleSolenoid::kReverse };
+
+	void SetLifter(LifterValue value);
+
+	bool IsLifterHigh();
```
Comment (duane701) [inline src/Subsystems/Shooter.h]: Change `IsLifterUp` to be `const`.

`bool IsLifterUp() const;`

File: src/Subsystems/Shooter.cpp (hunk 5 above)
Comment (duane701) [inline src/Subsystems/Shooter.cpp]: Make `IsLifterUp`, `const`. 

`bool Shooter::IsLifterUp() const`

File: src/Commands/PrepShooter.cpp [hunk 8]
```diff
@@ -0,0 +1,29 @@
+#include "PrepShooter.h"
+#include "SetLifter.h"
+#include "ShooterControl.h"
+
+PrepShooter::PrepShooter(double mSpeed)
+{
+	speed = mSpeed;
```
Comment (duane701) [inline src/Commands/PrepShooter.cpp]: Exchange `speed` and `mSpeed`. The `m` prefix is used to indicate a member variable. Here, `mSpeed` is the parameter.

Comment (duane701) [reply src/Commands/PrepShooter.cpp]: Use an initializer list:

```
PrepShooter::PrepShooter(double speed)
  : mSpeed(speed)
{
```

File: src/Commands/PrepShooter.h [hunk 9]
```diff
@@ -0,0 +1,16 @@
+#ifndef PrepShooter_H
+#define PrepShooter_H
+
+#include "Commands/CommandGroup.h"
+#include "WPILib.h"
+
+class PrepShooter: public CommandGroup
+{
+public:
+	PrepShooter(double mSpeed);
```
Comment (duane701) [inline src/Commands/PrepShooter.h]: Rename `mSpeed` to `speed`.

File: src/Commands/PrepShooter.h [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+private:
+	double speed;
```
Comment (duane701) [inline src/Commands/PrepShooter.h]: Rename `speed` to `mSpeed`. The `m` prefix signifies a member variable.

File: src/Commands/SetLifter.cpp [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+void SetLifter::Initialize()
+{
+
+}
+
+// Called repeatedly when this Command is scheduled to run
+void SetLifter::Execute()
+{
+Shooter::LifterValue
+	= Shooter::kUp ? Robot::shooter->SetLifter(Shooter::kUp) : Robot::shooter->SetLifter(Shooter::kUp);
```
Comment (duane701) [inline src/Commands/SetLifter.cpp]: No need to evaluate the value here. You're passing the value with the same type straight to the shooter.  Just use

`Robot::shooter(mShooterValue);`

File: src/Commands/SetLifter.cpp [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+//	Robot::shooter->SetLifter(Shooter::LifterValue)
+}
+
+// Make this return true when this Command no longer needs to run execute()
+bool SetLifter::IsFinished()
+{
+	return false;
```
Comment (duane701) [inline src/Commands/SetLifter.cpp]: Just `return true` for a solenoid. There's no point in continuously setting.

File: src/Commands/SetLifter.cpp [hunk 13]
```diff
@@ -0,0 +1,41 @@
+#include "SetLifter.h"
+
+SetLifter::SetLifter(Shooter::LifterValue)
```
Comment (duane701) [inline src/Commands/SetLifter.cpp]: Need  member variable to save the value. e.g. `mShooterValue`.

File: src/Commands/ShooterControl.cpp [hunk 14]
```diff
@@ -0,0 +1,39 @@
+#include "ShooterControl.h"
+
+ShooterControl::ShooterControl(double mSpeed)
```
Comment (duane701) [inline src/Commands/ShooterControl.cpp]: Rename `mSpeed` to `speed`.

File: src/Commands/ShooterControl.cpp [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+{
+	speed = mSpeed;
```
Comment (duane701) [inline src/Commands/ShooterControl.cpp]: Swap `speed` and `mSpeed`.

File: src/Commands/ToggleLifter.cpp [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+{
+
+}
+
+// Called repeatedly when this Command is scheduled to run
+void ToggleLifter::Execute()
+{
+Shooter::LifterValue value
+	= Robot::shooter->IsLifterUp() ? Shooter::kDown: Shooter::kUp;
+Robot::shooter->SetLifter(value);
```
Comment (duane701) [inline src/Commands/ToggleLifter.cpp:21]: Whitespace indent.

**Review by duane701 — COMMENTED** [review]
Only a few comments.

**Comment by duane701** [comment]
@OliviaRS -- Comments are for you.

File: src/Subsystems/Shooter.cpp (hunk 4 above)
Comment (OliviaRS) [reply src/Subsystems/Shooter.cpp]: Do you mean the auto-generated comment?

File: src/Commands/SetLifter.cpp (hunk 13 above)
Comment (duane701) [reply src/Commands/SetLifter.cpp]: Expanding my comment:

`SetLifter::SetLifter(Shooter::LifterValue lifterValue)`

File: src/Commands/SetLifter.cpp [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+{
+mLifterValue = Shooter::LifterValue;
```
Comment (duane701) [inline src/Commands/SetLifter.cpp]: `Shooter::LifterValue` is a type. You need a parameter name. Above I create a parameter name `lifterValue`. This line now becomes:

`mLifterValue = lifterValue;`

Better would be to use an initializer list:

```
SetLifter::SetLifter(Shooter::LifterValue lifterValue)
: mLifterValue(lifterValue)
{
}
```

**Comment by duanem** [comment]
If this is all already done we can close this PR.

**Comment by duane701** [comment]
Closing in favor of #6.

**Pull request closed without merging**